### PR TITLE
cudnn batch norm

### PR DESCRIPTION
**Pull request by piiswrong**

### CODE REVIEW

**Comment by tqchen** [comment]
Just want  to confirm if CuDNN update the moving statistics in forward instead of backward, this might affect the advanced mode for the memory optimization, so maybe we want to disable it for now if there is no significant speed gain.

**Comment by piiswrong** [comment]
Currently it's registered as a separate op anyway because it's not compatible with the original interface.

**Comment by antinucleon** [comment]
@piiswrong In the advanced mode, bn forward will be called many times. BTW, I notice CuDNN Activation costs more memory and a little bit slower. We may consider a better way to make choice of CuDNN or our implementation.

**Comment by piiswrong** [comment]
Ok I get what you mean. But how can it cost more memory? It doesn't need out_affine, so should save half of the memory right?

**Comment by antinucleon** [comment]
Dropout has been added in blacklist, I think you need to add cudnn bn here https://github.com/dmlc/mxnet/blob/master/src/symbol/static_graph.cc#L255

**Comment by tqchen** [comment]
Also mention the advanced mode is still experimental and disabled by default

**Comment by antinucleon** [comment]
@tqchen Let's merge it first, then add into blacklist?

**Comment by tqchen** [comment]
OK, if the review passes, and it works for the testcases by replacing batchnorm let us merge it in

File: src/operator/cudnn_batch_norm-inl.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  int NumOutputs() const override {
+    return 3;
+  }
+
+  std::vector<std::string> ListArguments() const override {
+    return {"data", "gamma", "beta"};
+  }
+
+  std::vector<std::string> ListOutputs() const override {
+    return {"output", "mean", "inv_var"};
```
Comment (antinucleon) [inline src/operator/cudnn_batch_norm-inl.h]: Can we change ```inv_var``` to same name as original batch norm implementation?

File: src/operator/cudnn_batch_norm-inl.h [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  std::vector<std::string> ListAuxiliaryStates() const override {
+    return {"moving_mean", "moving_inv_var"};
```
Comment (antinucleon) [inline src/operator/cudnn_batch_norm-inl.h]: same

File: src/operator/cudnn_batch_norm-inl.h (hunk 1 above)
Comment (piiswrong) [inline src/operator/cudnn_batch_norm-inl.h]: I renamed it because cudnn batch norm use 1/var instead of var. It shouldn't be loaded into batchnorm
